language/move-prover/stackless-bytecode-generator/src/stackless_bytecode.rs

```diff
@@ -78,69 +107,144 @@ pub enum BranchCond {
+    Load(AttrId, TempIndex, Constant),
+    Unary(AttrId, UnaryOp, TempIndex, TempIndex),
+    Binary(AttrId, BinaryOp, TempIndex, TempIndex, TempIndex),
 
-    Branch(CodeOffset, BranchCond),
-    Abort(TempIndex),
-    Destroy(TempIndex),
+    Branch(AttrId, Label, BranchCond),
+    Labeled(AttrId, Label, Box<Bytecode>),
+    Abort(AttrId, TempIndex),
+    Destroy(AttrId, TempIndex),
+    Nop(AttrId),
 }
 
 impl Bytecode {
+    pub fn get_attr_id(&self) -> AttrId {
+        use Bytecode::*;
+        match self {
+            Assign(id, ..)
+            | ReadRef(id, ..)
+            | WriteRef(id, ..)
+            | FreezeRef(id, ..)
+            | Call(id, ..)
+            | Ret(id, ..)
+            | Pack(id, ..)
+            | Unpack(id, ..)
+            | BorrowLoc(id, ..)
+            | BorrowField(id, ..)
+            | BorrowGlobal(id, ..)
+            | MoveToSender(id, ..)
+            | MoveFrom(id, ..)
+            | Exists(id, ..)
+            | Load(id, ..)
+            | Unary(id, ..)
+            | Binary(id, ..)
+            | Branch(id, ..)
+            | Labeled(id, ..)
+            | Abort(id, ..)
+            | Destroy(id, ..)
+            | Nop(id) => *id,
+        }
+    }
+
     pub fn is_unconditional_branch(&self) -> bool {
         matches!(
             self,
-            Bytecode::Ret(_) | Bytecode::Abort(_) | Bytecode::Branch(_, BranchCond::Always)
+            Bytecode::Ret(..) | Bytecode::Abort(..) | Bytecode::Branch(_, _, BranchCond::Always)
         )
     }
 
     pub fn is_conditional_branch(&self) -> bool {
         matches!(
             self,
-            Bytecode::Branch(_, BranchCond::False(_)) | Bytecode::Branch(_, BranchCond::True(_))
+            Bytecode::Branch(_, _, BranchCond::False(_)) | Bytecode::Branch(_, _, BranchCond::True(_))
         )
     }
 
     pub fn is_branch(&self) -> bool {
         self.is_conditional_branch() || self.is_unconditional_branch()
     }
 
+    pub fn skip_labelled(&self) -> &Bytecode {
+        if let Bytecode::Labeled(_, _, code) = self {
+            code
+        } else {
+            self
+        }
+    }
+
     /// Return the destination of branching if self is a branching instruction
-    pub fn branch_dest(&self) -> Option<&CodeOffset> {
+    pub fn branch_dest(&self) -> Option<Label> {
         match self {
-            Bytecode::Branch(offset, _) => Some(offset),
+            Bytecode::Branch(_, label, _) => Some(*label),
             _ => None,
         }
     }
 
-    /// Return the successor offsets of this instruction
-    pub fn get_successors(pc: CodeOffset, code: &[Bytecode]) -> Vec<CodeOffset> {
+    /// Returns a mapping from labels to code offsets.
+    pub fn label_offsets(code: &[Bytecode]) -> BTreeMap<Label, CodeOffset> {
+        let mut res = BTreeMap::new();
+        for (offs, bc) in code.iter().enumerate() {
+            if let Bytecode::Labeled(_, label, ..) = bc {
+                res.insert(*label, offs as CodeOffset);
```

Comment:
Is  it possible for there to be an existing entry for `label` here?

Reply:
Not for well-formed code. Added an assert to reflect this.